src/pam/main.cc

```diff
@@ -0,0 +1,422 @@
+  ConfirmationType confirmation_type(ConfirmationType::Unset);
+
+  // This task wait for the status of the python subprocess (we don't want a
+  // zombie process)
+  optional_task<int> child_task([&] {
+    int status;
+    wait(&status);
+    {
+      std::unique_lock<std::mutex> lk(m);
+      if (confirmation_type == ConfirmationType::Unset) {
+        confirmation_type = ConfirmationType::Howdy;
+      }
+    }
+    cv.notify_one();
+
+    return status;
+  });
+  child_task.activate();
+
+  // This task waits for the password input (if the workaround wants it)
+  optional_task<std::tuple<int, char *>> pass_task([&] {
+    char *auth_tok_ptr = nullptr;
+    int pam_res = pam_get_authtok(
+        pamh, PAM_AUTHTOK, const_cast<const char **>(&auth_tok_ptr), nullptr);
+    {
+      std::unique_lock<std::mutex> lk(m);
+      if (confirmation_type == ConfirmationType::Unset) {
+        confirmation_type = ConfirmationType::Pam;
+      }
+    }
+    cv.notify_one();
+
+    return std::tuple<int, char *>(pam_res, auth_tok_ptr);
+  });
+
+  if (auth_tok) {
+    pass_task.activate();
+  }
+
+  // Wait for the end either of the child or the password input
+  {
+    std::unique_lock<std::mutex> lk(m);
+    cv.wait(lk, [&] { return confirmation_type != ConfirmationType::Unset; });
+  }
+
+  // The password has been entered or an error has occurred
+  if (confirmation_type == ConfirmationType::Pam) {
+    // We kill the child because we don't need its result
+    kill(child_pid, SIGTERM);
+    child_task.stop(false);
+
+    // We just wait for the thread to stop since it's this one which sent us the
+    // confirmation type
+    pass_task.stop(false);
+
+    char *password = nullptr;
+    std::tie(pam_res, password) = pass_task.get();
+
+    if (pam_res != PAM_SUCCESS) {
+      return pam_res;
+    }
+
+    // The password has been entered, we are passing it to PAM stack
+    return PAM_IGNORE;
+  }
+
+  // The compare process has finished its execution
+  child_task.stop(false);
+
+  // We want to stop the password prompt, either by canceling the thread when
+  // workaround is set to "native", or by emulating "Enter" input with
+  // "input"
+
+  // UNSAFE: We cancel the thread using pthread, pam_get_authtok seems to be
+  // a cancellation point
+  if (workaround == Workaround::Native && pass_task.is_active()) {
+    pass_task.stop(true);
+  } else if (workaround == Workaround::Input) {
+    if (geteuid() != 0) {
+      syslog(LOG_WARNING, "Insufficient permission to create the fake device");
+      conv_function(PAM_ERROR_MSG,
+                    S("Insufficient permission to send Enter "
+                      "press, waiting for user to press it instead"));
+    } else {
+      try {
+        EnterDevice enter_device;
+        for (int retries = 0;
+             retries < MAX_RETRIES &&
+             pass_task.wait(DEFAULT_TIMEOUT) == std::future_status::timeout;
+             retries++) {
+          enter_device.send_enter_press();
+        }
```

Comment:
Maybe add a log message for when max retries are exceeded?

Reply:
Thanks, I added your suggestion!